### ISSUE TITLE
Deploy seed nodes to assist p2p network discovery

### ISSUE THREAD

**Issue by ghost** [bisq-network/bisq#3]
## Motivation

Server-side seed nodes are necessary such that that first-time Bitsquare users can "bootstrap" into the p2p network, i.e. discover enough other nodes and communicate with them. From that point forward, clients maintain their own list of last-known-good peers, meaning that clients no longer rely on the seed nodes. 
## Next steps:
- [x] Steve to merge seed node command line startup from the seedpeer branch (eaaa0e0)
- [x] Steve and Manfred to get their local routers working with UPnP and Bitsquare (per https://github.com/bitsquare/bitsquare/issues/3#issuecomment-58877139)
- [ ] Wait for Thomas Bocek to look into TomP2P issues and get back to us (per https://github.com/bitsquare/bitsquare/issues/3#issuecomment-58877139)

**Comment by bisq-network** [comment]
I tried again with the current master to test the serverside bootstrapping, but it still has the same issue that the relay fallback causes an exception on the server. Thomas Bocek told me on the mailing list that he will look into the issue. There is also a warning with the signing for data protection. That seems to be another open bug, he will investigate.
I added code that the last successful bootstrap method is saved and reused so a repeated connection in relay or port forwarding mode will be faster.

**Comment by notmandatory** [comment]
I'll investigate after I've got my akka prototype going.

**Comment by bisq-network** [comment]
Probably we need to wait until Thomas has fixed the problem...

cu,
manfred

Am 22.08.2014 um 20:51 schrieb stevemyers notifications@github.com:

**Comment by ManfredKarrer** [comment]
Tested latest master of TomP2P but does not solve any of the problems. Same errors and warnings. Reported it to Thomas Bocek: http://lists.tomp2p.net/pipermail/users/2014-October/000395.html

**Comment by cbeams** [comment]
@ManfredKarrer, @stevemyers: per our team call:
- We will focus on UPnP for auto net port config and manually configured router port forward

Steve or Manfred, please update with whatever next steps, and please add the `in progress` label if this is in fact in progress. Also please assign to yourself if you're actually on point now. Thanks.

**Comment by ManfredKarrer** [comment]
Update:
Local router is setup for supporting UPnP and port forwarding but problems are the same. Tested also with minimal setup code form NATExample.java source form TomP2P examples and run on server the serverside code from that but also without success. Tested also with Steves server, smae problem. I reported Thomas that we also cannot get a connection wiht his example code.
He is currently trying to get Bitsquare running and will investigate.

**Comment by cbeams** [comment]
Thanks, Manfred. I've reflected all of this in the "Next Steps" section in the description above. I've also added the `blocked` label to reflect that we're apparently stuck until Thomas gets back to us.

**Comment by ManfredKarrer** [comment]
I have removed the blocked label as we are continuing investigating in parallel. I will cheeck out other projects uning TomP2P and Steve might test other network settings. Maybe I will ask other team members to run the app against a server seed node to see if the problems are only with my network setup or for others as well.

**Comment by notmandatory** [comment]
One interesting thing I found when I was running the Bitsquare node in seed mode on my Digital Ocean server is it seems to be listening with IPv6 instead of IPv4 (I set it to listen on port 50010), see below.  I’m trying to get the BSExampleNATServer running on my Digital Ocean server now to see if it does the same thing.  If so this could be the reason remote peers can't connect to our DO seed nodes.

```
root@bitsquare:~/javapackager-app# netstat -an
Active Internet connections (servers and established)
Proto Recv-Q Send-Q Local Address           Foreign Address         State      
tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN     
tcp6       0      0 :::50010                :::*                    LISTEN     
tcp6       0      0 :::22                   :::*                    LISTEN     
udp6       0      0 :::50010                :::*          
```

**Comment by notmandatory** [comment]
I tried running the BSExampleNATServer and got the same results:

```
root@bitsquare:~# netstat -an
Active Internet connections (servers and established)
Proto Recv-Q Send-Q Local Address           Foreign Address         State      
tcp        0      0 0.0.0.0:22              0.0.0.0:*               LISTEN     
tcp6       0      0 :::5000                 :::*                    LISTEN     
tcp6       0      0 :::22                   :::*                    LISTEN     
udp6       0      0 :::5000                 :::*                         
```

**Comment by notmandatory** [comment]
Nevermind.  Even with binding set to INET the netstat on Digital Ocean still shows Proto tcp6, and with or without hardcoding to INET I see my client connecting on the Digital Ocean server, so the issue is still with the local UPnP not successfully setting up port forwarding.

**Comment by ManfredKarrer** [comment]
Do you think it might make sense to try out another VM on amazon or so?

**Comment by cbeams** [comment]
On Oct 13, 2014, at 6:17 PM, @ManfredKarrer wrote (via email):

> @bitsquare/contributors,
> 
> If you have time could you please update to lastest master and run `net.tomp2p.examples.BSExampleNAT` and send me the logs?
> 
> It should connect to the serverside bootstrapping seednode and I wanted to see if you also fall back to relay mode or if it is just on my network setup.

**Comment by cbeams** [comment]
> could you please update to lastest master and run `net.tomp2p.examples.BSExampleNAT` and send me the logs

Sure. Running against 8983ce1, I get the following:

```
10:46:24.615 [NETTY-TOMP2P - worker-client/server - -1-1] WARN  io.netty.channel.AbstractChannel - A non-root user can't receive a broadcast packet if the socket is not bound to a wildcard address; binding to a non-wildcard address (/192.168.0.69:5500) anyway as requested. 
10:46:24.619 [NETTY-TOMP2P - worker-client/server - -1-2] WARN  io.netty.channel.AbstractChannel - A non-root user can't receive a broadcast packet if the socket is not bound to a wildcard address; binding to a non-wildcard address (/127.0.0.1:5500) anyway as requested. 
failed Future (compl/canc):true/false, FAILED, FutureDiscover: We need at least the TCP connection <-> Future (compl/canc):true/false, FAILED, Channel creation failed java.net.ConnectException: Connection refused: /128.199.251.106:5000
10:46:25.253 [NETTY-TOMP2P - worker-client/server - -1-3] WARN  net.tomp2p.connection.Sender - Channel creation failed  java.net.ConnectException: Connection refused: /128.199.251.106:5000
    at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method) ~[na:1.8.0_05]
    at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:712) ~[na:1.8.0_05]
    at io.netty.channel.socket.nio.NioSocketChannel.doFinishConnect(NioSocketChannel.java:208) ~[netty-transport-4.0.21.Final.jar:4.0.21.Final]
    at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:287) ~[netty-transport-4.0.21.Final.jar:4.0.21.Final]
    at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:528) [netty-transport-4.0.21.Final.jar:4.0.21.Final]
    at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:468) [netty-transport-4.0.21.Final.jar:4.0.21.Final]
    at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:382) [netty-transport-4.0.21.Final.jar:4.0.21.Final]
    at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:354) [netty-transport-4.0.21.Final.jar:4.0.21.Final]
    at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:116) [netty-common-4.0.21.Final.jar:4.0.21.Final]
    at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:137) [netty-common-4.0.21.Final.jar:4.0.21.Final]
    at java.lang.Thread.run(Thread.java:745) [na:1.8.0_05]
```

**Comment by ManfredKarrer** [comment]
Ah sorry there was steves server set and it seemed that is not running. I just committed to use my server which is running now.
could you try again?
thanks!

Am 14.10.2014 um 10:53 schrieb Chris Beams notifications@github.com:

**Comment by cbeams** [comment]
(@stevemyers, see questions for you below)

Now against 5cb2c2c:

```
11:12:39.179 [NETTY-TOMP2P - worker-client/server - -1-3] WARN  net.tomp2p.nat.PeerNAT - cannot find UPNP devices 
11:12:46.951 [NETTY-TOMP2P - worker-client/server - -1-3] WARN  net.tomp2p.nat.PeerNAT - cannot find NAT-PMP devices 
failed Future (compl/canc):true/false, FAILED, We are most likely behind NAT, try to UPNP, NATPMP or relay {}, {}paddr[0x[/188.226.179.109,5000]
```

@ManfredKarrer and I are getting the same thing consistently (per discussions in #bitsquare channel).

@stevemyers, the result above was directly against my ISP-issued router (I usually have a Time Capsule in-between), and I explicitly enabled UPnP and remote config management. I believe you mentioned on our call on Sunday that you were looking at a standalone UPnP Java library? I don't see anything to that effect having made its way into master. Any update on that?

**Comment by ManfredKarrer** [comment]
It might be a problem on mac only? We all use macs...
@remyers Would you be able to test it on windows? I just have it on VM (i will test it there but dont expect it behaves like native).

**Comment by remyers** [comment]
I need someone to ELI5 how to run that tomp2p test.  I've done the
following:

C:\Users\Richard\Documents\GitHub\bitsquare> gradlew build

> BUILD SUCCESSFUL
> Total time: 1 mins 58.502 secs
> C:\Users\Richard\Documents\GitHub\bitsquare> java
> net.tomp2p.examples.BSExampleNAT
> and
> C:\Users\Richard\Documents\GitHub\bitsquare\build\classes\main\net\tomp2p\examples>
> java BSExampleNAT
> but I think there is some java thing I'm doing wrong because it always
> says:
> Error: Could not find or load main class .\BSExampleNAT

Suggestions?

**Comment by ManfredKarrer** [comment]
Hi Richard,

lets talk on irc...

br,
manfred

Am 14.10.2014 um 15:17 schrieb Richard Myers notifications@github.com:

**Comment by ManfredKarrer** [comment]
Update:

Chris tested with mac and got same result - relay fallback.

Richard tested on windows and he could not connect at all to the server, same was for me when testing with windows on my VM. On Linux VM i got elay again.
I deactivated the windows firewall but was same.

failed Future (compl/canc):true/false, FAILED, FutureDiscover: We need at least the TCP connection <-> Future (compl/canc):true/false, FAILED, Channel creation failed java.net.SocketException: Network is unreachable: no further information
paultroon
17:28:11.145 [NETTY-TOMP2P - worker-client/server - -1-3] WARN  net.tomp2p.connection.Sender - Channel creation failed  java.net.SocketException: Network is unreachable: no further information
paultroon
at sun.nio.ch.SocketChannelImpl.checkConnect(Native Method) ~[na:1.8.0_20]
paultroon
at sun.nio.ch.SocketChannelImpl.finishConnect(SocketChannelImpl.java:716) ~[na:1.8.0_20]
paultroon
at io.netty.channel.socket.nio.NioSocketChannel.doFinishConnect(NioSocketChannel.java:208) ~[netty-transport-4.0.21.Final.jar:4.0.21.Final]
paultroon
at io.netty.channel.nio.AbstractNioChannel$AbstractNioUnsafe.finishConnect(AbstractNioChannel.java:287) ~[netty-transport-4.0.21.Final.jar:4.0.21.Final]
paultroon
at io.netty.channel.nio.NioEventLoop.processSelectedKey(NioEventLoop.java:528) [netty-transport-4.0.21.Final.jar:4.0.21.Final]
paultroon
at io.netty.channel.nio.NioEventLoop.processSelectedKeysOptimized(NioEventLoop.java:468) [netty-transport-4.0.21.Final.jar:4.0.21.Final]
paultroon
at io.netty.channel.nio.NioEventLoop.processSelectedKeys(NioEventLoop.java:382) [netty-transport-4.0.21.Final.jar:4.0.21.Final]
paultroon
at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:354) [netty-transport-4.0.21.Final.jar:4.0.21.Final]
paultroon
at io.netty.util.concurrent.SingleThreadEventExecutor$2.run(SingleThreadEventExecutor.java:116) [netty-common-4.0.21.Final.jar:4.0.21.Final]
paultroon
at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:137) [netty-common-4.0.21.Final.jar:4.0.21.Final]
paultroon
at java.lang.Thread.run(Thread.java:745) [na:1.8.0_20]

I tested with a server and client from ExampleNAT from the old 4.4 release, but also no success, but different failure msg:
failed BaseFuture Status=(isComplete:true, reason:Timeout in Discover: 5010ms, type:FAILED)

I am wondering if that warning might be relevant to us?
18:40:49.385 [NETTY-TOMP2P - worker-client/server - -1-1] WARN  io.netty.channel.AbstractChannel - A non-root user can't receive a broadcast packet if the socket is not bound to a wildcard address; binding to a non-wildcard address (/192.168.1.33:5500) anyway as requested. 

Steve commented by mail already he thinks not.

**Comment by notmandatory** [comment]
@cbeams the UPnP library I mentioned on our last call is the library that is already integrated into TomP2P, it's called [Weupnp](http://bitletorg.github.io/weupnp/). I was just saying it seems to be better supported than the PMP support (for Apple devices)  which was created by Thomas Bocek himself.

**Comment by ManfredKarrer** [comment]
UPnP is used more on windows, right? what is used on unix normally? UPnP or PMP? we have not got successful connections from windows yet. so we dont knwo if it also falls back to replay. might be some firewall problem (i deactivated it on my VM, but not sure if the setup with the vm is a good test setup, might behave different then native).

**Comment by notmandatory** [comment]
NAT-PMP seems to be mainly an Apple (and some open source) supported standard. In Thomas' implementation it makes an external OS call to netstat to find the default router.

http://en.wikipedia.org/wiki/NAT_Port_Mapping_Protocol

UPnP is used for nearly all non-Apple routers and uses broadcast messages to find the router so unlike NAT-PMP it doesn't need to ask the OS what the default router IP is:

http://en.wikipedia.org/wiki/Universal_Plug_and_Play

**Comment by notmandatory** [comment]
I was able to capture TomP2P trying to setup UPnP port forwarding.  It looks like it set it up.. but the external port is different from the internal port.  I'm not sure if this is by design or not.  The port my client is using is 55010.

```
POST /ctl/IPConn HTTP/1.1
Content-Type: text/xml
SOAPAction: "urn:schemas-upnp-org:service:WANIPConnection:1#AddPortMapping"
Connection: Close
User-Agent: Java/1.8.0_20
Host: 192.168.0.1:5000
Accept: text/html, image/gif, image/jpeg, *; q=.2, */*; q=.2
Content-Length: 641
```

``` XML
<?xml version="1.0"?>
<SOAP-ENV:Envelope xmlns:SOAP-ENV="http://schemas.xmlsoap.org/soap/envelope/" SOAP-ENV:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/">
<SOAP-ENV:Body>
<m:AddPortMapping xmlns:m="urn:schemas-upnp-org:service:WANIPConnection:1"><NewRemoteHost></NewRemoteHost>
<NewInternalClient>192.168.0.7</NewInternalClient>
<NewProtocol>TCP</NewProtocol>
<NewEnabled>1</NewEnabled>
<NewLeaseDuration>0</NewLeaseDuration>
<NewInternalPort>55010</NewInternalPort>
<NewExternalPort>54880</NewExternalPort>
<NewPortMappingDescription>TomP2P mapping TCP</NewPortMappingDescription></m:AddPortMapping></SOAP-ENV:Body>
</SOAP-ENV:Envelope>
```

```
HTTP/1.1  200 OK
Content-Type: text/xml
Connection: close
Content-Length: 263
Server: Arris/1.0 UPnP/1.0 miniupnpd/1.0
```

``` xml
<?xml version="1.0"?>
<s:Envelope xmlns:s="http://schemas.xmlsoap.org/soap/envelope/" s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/">
<s:Body>
<u:AddPortMappingResponse xmlns:u="urn:schemas-upnp-org:service:WANIPConnection:1"/>
</s:Body>
</s:Envelope>
```

**Comment by ManfredKarrer** [comment]
There have been some discussion from Pawal on the mailing list regarding external port. I tried that already but it has not changed anything on my side.

 Peer peer = new PeerBuilder(keyPair).ports(port).portsExternal(port)
                    .channelServerConfiguration(csc).start();

For timeout problems I read somewhere in the list to increase the timeout at ChannelServerConficuration. Tried that as well but no success as well. 

ChannelServerConficuration csc = PeerBuilder.createDefaultChannelServerConfiguration();
            csc.idleTCPSeconds(20).idleUDPSeconds(20).connectionTimeoutTCPMillis(20000);
            Peer peer = new PeerBuilder(keyPair).ports(randomPort).channelServerConfiguration(csc).start();

br,
Manfred

Am 15.10.2014 um 07:24 schrieb Steve Myers notifications@github.com:

**Comment by ManfredKarrer** [comment]
GOOD NEWS!
I got a non-relay connection working when setting up manual port forwarding (portsExternal is needed to be set to the client port). So it seems that PMP (on OSX) is not supported correctly. 
I got still not the trade process running and the DHT storage seems to be pretty slow, but will investigate further, maybe we need to tweak some timouts or other configs.

http://lists.tomp2p.net/pipermail/users/2014-October/000405.html

**Comment by notmandatory** [comment]
That's great! Unless you're using an Airport I suspect PMP isn't supported by your Inet router.

**Comment by ManfredKarrer** [comment]
Update: 
Got some help form Thomas and Nico who is involved in TomP2P. DHT storage seems to work now reliable. I got today a laptop from a friend from the meetup (Stan) as contribution ot the project and will start to test tomorrow on windows and linux. Direct messaging (trade process) still makes problems but we are getting closer.

**Comment by ManfredKarrer** [comment]
Good news from Thomas:

Hi Manfred,
Me and Jonas tested the UPNP and added some fixes. We also tested with
MAC, so UPNP should work now.
Regards,
Thomas

I tested the actual head and UPnP is not working. Not without strange troubles first...
It seemes that there is a bug in the parser in the weupnp library when there is not the actual time set in the router. Timeserver was disabled in the router and first the UPnP did nto succeed, then i enabled timeserver and UPnP worked afterwards. Not absolutely sure if it the reason, as I cannot set back the time (maybe a reset would do it, but dont want any more troubles yet...) to reproduce it.
Anyway, hopefully most other routers have a better setup, and maybe we can report/fix later the parsing issue in the weupnp library. 
I will go on testing to see how much has been improved with that fix in the TomP2P lib. If tests are positive I will deploy a new snapshot and update our fork to it.

**Comment by cbeams** [comment]
Let me know if you need help deploying the new snapshot.

**Comment by ManfredKarrer** [comment]
@cbeams Thanks Chris! I will try it on my own first, if I have troubles I will ping you...

**Comment by ManfredKarrer** [comment]
Good news from Thomas, he said we think he can fix the open problem with directSend (UPnP) until mid of the week.

**Comment by ManfredKarrer** [comment]
@stevemyers @cbeams @remyers 

Hi Steve,

Great!
Is directSend working for you?

Good to have the master seeds connected. I will add a test where one peer connected to my server stores data and the 2. peer connected to your server read the data.
Will check afterwards to get the offer storage working again in the app and to change back to async handling so the app does not freeze during startup.

Maybe we can extend later the tests in the way that we have a number of normal peers (5-10) running on the server and then check if shutdown of nodes will affect the DHT storage. It should automatically manage the storages, so a leaving node will be replaced by others. As long 1 node remains the data should be accessible.
Would be cool if we could manage to get a test environment (#51) which runs those scenarios automatically. Will be a bit complex though, but I think we should have something in the WAN for testing real life conditions.
Maybe there are some test cases in the tomp2p lib we can extend. There is also a sim grid (https://github.com/tomp2p/TomP2P-SimGrid) test setup. I have not looked into that yet, but I think it is only for local testing of massive nodes (10 000s). Would be more interesting to get something like that for the WAN.
But of course that complex test environment is no hi prio yet.

Do you have any ideas how we should go on with the directSend problem? I hope Thomas can fix it soon, he wrote that he thinks he can fix it until mid of the week, so chances are good that we get it soon. Otherwise probably its worth to dive deeper into tomp2p code, will probably be not the last problem and better to be not too dependent on the help of Thomas. Though that will be a time consuming adventure... but I think we need to go that path. What do you think?

**Comment by ManfredKarrer** [comment]
Good news:
Direct send and DHT in relay mode is now working in the latest head of the master (just tested in our test class). Unfortunately I dont get port forwarding anymore. Will investigate further...

**Comment by ManfredKarrer** [comment]
Trade process with relay mode works now. Offer storage + messaging!
Will continue to find out why port forwarding does not work anymore. And there is a problem with offer removal. I will wait update our fork soon.

**Comment by ManfredKarrer** [comment]
Added tests for add (used for offers to add an offer to a list), remove and all tests are green with relay. There are 2 small issues with incorreect success results, whcih i will report Thomas. I will update the our fork to the master head.

**Comment by ManfredKarrer** [comment]
Port forwarding is working again, dont know why it was not working before. Maybe my router behaves strange with UPnP? But there are still problems, DHT basically works but delivers wrong success flags, and directSend is still not working. If we dont get it working we can deactivate it so users can only connect directly or via relay mode (just as temporary solution).

**Comment by ManfredKarrer** [comment]
I deactivated port forwarding for the moment as that is not working correctly. In relay mode startup is slower but all works so far as I have tested.
